BLUESPAWN-client/src/util/pe/Import_Section.cpp

```diff
@@ -0,0 +1,145 @@
+	DWORD dwNameOffset = image.expanded ? ImportDirectoryTable.dwNameRVA : image.RVAToOffset(ImportDirectoryTable.dwNameRVA);
+	DWORD dwLookupTableOffset = image.expanded ? ImportDirectoryTable.dwImportLookupTableRVA : image.RVAToOffset(ImportDirectoryTable.dwImportLookupTableRVA);
+	DWORD dwIATOffset = image.expanded ? ImportDirectoryTable.dwImportAddressTableRVA : image.RVAToOffset(ImportDirectoryTable.dwImportAddressTableRVA);;
+	this->sLibraryName = StringToWidestring(image.base.GetOffset(dwNameOffset).ReadString());
+	if(image.arch == x64){
+		MemoryWrapper<IMPORT_LOOKUP_TABLE64> entry = image.base.GetOffset(dwLookupTableOffset).Convert<IMPORT_LOOKUP_TABLE64>();
+		while(entry->value){
+			Import i = {};
+			if(entry->type){
+				i = { true, static_cast<WORD>(entry->value), 0 };
+			} else {
+				DWORD dwHintOffset = image.expanded ? entry->value : image.RVAToOffset(entry->value);
+				WORD hint = *image.base.GetOffset(dwHintOffset).Convert<WORD>();
+				std::string name = image.base.GetOffset(dwHintOffset + 2).ReadString();
+				i = { false, 0, {hint, name} };
+			}
+			vImportToIAT.emplace_back(std::pair<Import, DWORD>{ i, dwIATOffset });
+			dwLookupTableOffset += 8;
+			dwIATOffset += 8;
+		}
+	} else {
+		MemoryWrapper<IMPORT_LOOKUP_TABLE32> entry = image.base.GetOffset(dwLookupTableOffset).Convert<IMPORT_LOOKUP_TABLE32>();
+		while(entry->value){
+			Import i = {};
+			if(entry->type){
+				i = { true, static_cast<WORD>(entry->value), 0 };
+			} else {
+				DWORD dwHintOffset = image.expanded ? entry->value : image.RVAToOffset(entry->value);
+				WORD hint = *image.base.GetOffset(dwHintOffset).Convert<WORD>();
+				std::string name = image.base.GetOffset(dwHintOffset + 2).ReadString();
+				i = { false, 0, {hint, name} };
+			}
+			vImportToIAT.emplace_back(std::pair<Import, DWORD>{ i, dwIATOffset });
+			dwLookupTableOffset += 4;
+			dwIATOffset += 4;
+		}
+	}
+}
+
+bool Import_Library::LoadImportLibrary(const HandleWrapper& context){
+	auto info = Image_Loader(context).GetImageInfo(sLibraryName);
+	if(!info.has_value()){
+		return false;
+	}
+
+	auto& TargetImage = info->GetImage();
+	for(auto pair : vImportToIAT){
+		auto im = pair.first;
+		auto offset = pair.second;
+		if(im.IsOrdinal){
+			if(image.arch == x64){
+				DWORD64 address = TargetImage.exports->GetExportAddress(im.ordinal);
+				if(!address){
+					return false;
+				}
+
+				image.base.Write(reinterpret_cast<CHAR*>(&address), sizeof(address), offset);
+			}
+		}
+	}
+	return true;
+}
+
+Import_Section::Import_Section(const PE_Section& section) : PE_Section{ section } {
+	MemoryWrapper<IMPORT_DIRECTORY_TABLE> table = SectionContent.Convert<IMPORT_DIRECTORY_TABLE>();
+	while((*table).dwImportLookupTableRVA){
+		imports.emplace_back(Import_Library{ section.AssociatedImage, *table });
+	}
+}
+
+bool Import_Section::LoadAllImports(const HandleWrapper& context){
+	std::set<std::wstring> images = {};
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'images' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=4994516)